storage/rocksdb/rdb_datadic.cc

```diff
@@ -925,13 +988,67 @@ uint Rdb_key_def::pack_record(const TABLE *const tbl, uchar *const pack_buffer,
     tuple = pack_field(field, &m_pack_info[i], tuple, packed_tuple, pack_buffer,
                        unpack_info, n_null_fields);
 
+    // Check if this column is covered.
+    if (unpack_info && m_index_type == INDEX_TYPE_SECONDARY) {
+      size_t data_length = field->data_length();
+      uint16 key_length;
+
+      // We always need the PK as a suffix, but in some cases MySQL won't add
+      // the PK suffix to a secondary key, and therefore it won't be included
+      // in key_info. If so, we need to get the key length for key parts from
+      // the PK directly.
+      DBUG_ASSERT(tbl->key_info != nullptr);
+      if (i >= tbl->key_info[get_keyno()].actual_key_parts) {
+        int j = i - tbl->key_info[get_keyno()].actual_key_parts;
+        key_length = tbl->key_info[tbl->s->primary_key].key_part[j].length;
+      } else {
+        key_length = tbl->key_info[get_keyno()].key_part[i].length;
+      }
+
+      enum_field_types type = field->real_type();
+      // Can't do index-only scans on strings without an unpack function.
+      const bool string_type =
+          (type == MYSQL_TYPE_VARCHAR) || (type == MYSQL_TYPE_STRING);
+      const bool can_unpack = !string_type || m_pack_info[i].m_unpack_func;
+      if (can_unpack && data_length <= key_length) {
+        // Currently we only support index-only scans with these types.
+        // Warning: Rdb_field_packing::setup should only return true for types
+        // which are in this list.
+        switch (type) {
+        case MYSQL_TYPE_LONGLONG:
+        case MYSQL_TYPE_LONG:
+        case MYSQL_TYPE_INT24:
+        case MYSQL_TYPE_SHORT:
+        case MYSQL_TYPE_TINY:
+        case MYSQL_TYPE_DOUBLE:
+        case MYSQL_TYPE_FLOAT:
+        case MYSQL_TYPE_NEWDECIMAL:
+        case MYSQL_TYPE_DATETIME2:
+        case MYSQL_TYPE_TIMESTAMP2:
+        case MYSQL_TYPE_TIME2:
+        case MYSQL_TYPE_YEAR:
+        case MYSQL_TYPE_NEWDATE:
+        case MYSQL_TYPE_VARCHAR:
+        case MYSQL_TYPE_STRING:
+          bitmap_set_bit(&covered_bitmap, i);
+          break;
+        default:
+          break;
+        }
+      }
+    }
+
     // Restore field->ptr and field->null_ptr
     field->move_field(tbl->record[0] + field_offset,
                       maybe_null ? tbl->record[0] + null_offset : nullptr,
                       field->null_bit);
   }
 
   if (unpack_info) {
+    if (m_index_type == INDEX_TYPE_SECONDARY) {
+      unpack_info->write_uint16_at(covered_bitmap_pos, covered_bits);
```

Comment:
There is an optimization below that tries not to store unpack_info if none of the columns requires it. This always adds an extra 2 bytes to unpack info, which means len != RDB_UNPACK_HEADER_SIZE, so this means we always write out unpack_info for secondary indexes, even if it's not needed (e.g. secondary index consists entire of ints).